mesonbuild/dependencies/mpi.py

```diff
@@ -244,3 +247,53 @@ def __init__(self, name: str, env: 'Environment', kwargs: T.Dict[str, T.Any],
         self.compile_args = ['-I' + incdir, '-I' + os.path.join(incdir, post)]
         if self.language == 'fortran':
             self.link_args.append('-l' + os.path.join(libdir, 'msmpifec'))
+
+
+class IMPIDependency(SystemDependency):
+
+    """Intel(R) MPI for Windows."""
+
+    def __init__(self, name: str, env: 'Environment', kwargs: T.Dict[str, T.Any],
+                 language: T.Optional[str] = None):
+        super().__init__(name, env, kwargs, language=language)
+        # only for windows
+        if not self.env.machines[self.for_machine].is_windows():
+            return
+        # only for x86 and x64
+        arch = detect_cpu_family(self.env.coredata.compilers.host)
+        if arch not in ('x86', 'x86_64'):
+            self.is_found = False  # TODO: is this line correct?
+            return
+
+        rootdir = os.environ.get('I_MPI_ROOT')
+        if rootdir is None:
+            self.is_found = False
+            return
+
+        subdir = 'ia32' if arch == 'x86' else 'intel64'
+        prefix = os.path.join(rootdir, subdir)
+        if not os.path.isdir(prefix):
+            prefix = rootdir
+
+        incdir = os.path.join(prefix, 'include')
+        libdir = os.path.join(prefix, 'lib')
+
+        buildtype = env.coredata.optstore.get_value_for(OptionKey('buildtype'))
```

Comment:
This block may look a bit convoluted. The intention here is to cope with various install directory layouts that Intel MPI has used across releases and distribution methods (offline installer, nuget, pip/conda packages).